src/server/repositories/UrlRepository.ts

```diff
@@ -0,0 +1,240 @@
+import { inject, injectable } from 'inversify'
+import { BaseRepository } from './BaseRepository'
+import { Url, UrlType } from '../models/url'
+import { NotFoundError } from '../util/error'
+import { redirectClient } from '../redis'
+import { logger, redirectExpiry } from '../config'
+import { transaction } from '../util/sequelize'
+import DependencyIds from '../constants'
+import { FileVisibility, S3Interface } from '../util/aws'
+import { ACTIVE } from '../models/types'
+
+const { Public, Private } = FileVisibility
+
+/**
+ * A url repository that handles access to the data store of Urls.
+ */
+@injectable()
+export abstract class AbstractUrlRepository
+  implements BaseRepository<StorableUrl> {
+  abstract findOne(where: object): Promise<StorableUrl | null>
+
+  /**
+   * Updates the input url with the input changes and file (if any) in the data store.
+   * @param  {StorableUrl} url The url to modify
+   * @param  {object} changes The key value pairs contained in this object will overwrite those in the original url
+   * @param  {StorableFile} file? The file to change to (if any)
+   * @returns Promise
+   */
+  abstract update(
+    url: StorableUrl,
+    changes: object,
+    file?: StorableFile,
+  ): Promise<StorableUrl>
+
+  /**
+   * Create a new Url in the data store.
+   * @param  {{userId:number;shortUrl:string;longUrl?:string}} properties of new Url
+   * @param  {StorableFile} file? that this Url leads to, if any
+   * @returns Promise that resolves to the newly created url
+   */
+  abstract create(
+    properties: { userId: number; shortUrl: string; longUrl?: string },
+    file?: StorableFile,
+  ): Promise<StorableUrl>
+
+  /**
+   * Looks up the longUrl given a shortUrl from the cache, falling back
+   * to the database. The cache is re-populated if the database lookup is
+   * performed successfully.
+   * @param {string} shortUrl
+   * @returns {Promise<string>}
+   * @throws {NotFoundError}
+   */
+  abstract getLongUrl: (shortUrl: string) => Promise<string>
+
+  /**
+   * Asynchronously increment the number of clicks in the database.
+   * @param shortUrl
+   */
+  abstract incrementClick: (shortUrl: string) => Promise<void>
+}
+
+/**
+ * A url repository that handles access to the data store of Urls.
+ * The following implementation uses Sequelize, AWS S3 and Redis.
+ */
+export class UrlRepository extends AbstractUrlRepository {
+  private fileBucket: S3Interface
+
+  public constructor(@inject(DependencyIds.s3) fileBucket: S3Interface) {
+    super()
+    this.fileBucket = fileBucket
+  }
+
+  findOne: (where: object) => Promise<StorableUrl | null> = async (where) => {
+    return (await Url.findOne({ where: { ...where } })) as StorableUrl | null
```

Comment:
```suggestion
    return (await Url.findOne({ where })) as StorableUrl | null
```

Reply:
Typescript complains that it does not match Sequelize's types when it isn't spread for some reason.

Reply:
Found a way to typecast it to the sequelize type